TaffoInitializer/IndirectCallPatcher.cpp

```diff
@@ -0,0 +1,207 @@
+  copy(curCall->arg_begin() + 3, curCall->arg_end(),
+       back_inserter(trampolineArgs));
+
+  // Keep ref to the indirect function, preventing globaldce pass to destroy it
+  auto magicBitCast =
+      new BitCastInst(indirectFunction, indirectFunction->getType(), "", block);
+  ReturnInst::Create(m.getContext(), nullptr, block);
+
+  // Create the arguments of the direct function, extracted from the indirect
+  std::vector<Value *> outlinedArgumentsInsideTrampoline;
+  // Create null pointer to patch the internal OpenMP argument
+  Value *nullPointer = ConstantPointerNull::get(
+      PointerType::get(Type::getInt32Ty(trampolineFunction->getContext()), 0));
+  outlinedArgumentsInsideTrampoline.push_back(nullPointer);
+  outlinedArgumentsInsideTrampoline.push_back(nullPointer);
+  for (auto argIt = trampolineFunction->arg_begin() + 2;
+       argIt < trampolineFunction->arg_end(); argIt++) {
+    outlinedArgumentsInsideTrampoline.push_back(argIt);
+  }
+
+  // Create the call to the direct function inside the trampoline
+  CallInst *outlinedCall =
+      CallInst::Create(microTaskFunction, outlinedArgumentsInsideTrampoline);
+  outlinedCall->insertAfter(magicBitCast);
+
+  // Create the call to the trampoline function after the indirect function
+  CallInst *trampolineCallInstruction =
+      CallInst::Create(trampolineFunction, trampolineArgs);
+  trampolineCallInstruction->setCallingConv(
+      curCallInstruction->getCallingConv());
+  trampolineCallInstruction->insertBefore(curCallInstruction);
+  trampolineCallInstruction->setDebugLoc(curCallInstruction->getDebugLoc());
+
+  MDNode *indirectFunctionRef = MDNode::get(
+      curCallInstruction->getContext(), ValueAsMetadata::get(indirectFunction));
+  trampolineCallInstruction->setMetadata(INDIRECT_METADATA,
+                                         indirectFunctionRef);
+
+  // Save the old instruction to delete it later
+  toDelete.push_back(curCallInstruction);
+  LLVM_DEBUG(dbgs() << "Newly created instruction: "
+                    << *trampolineCallInstruction << "\n");
+}
+
+/// Map to keep track of the handled indirect functions.
+const std::map<const std::string, handler_function> indirectCallFunctions = {
+    {"__kmpc_fork_call", &handleKmpcFork},
+};
+
+/// Check if the given call is indirect, using the related handler.
+void handleIndirectCall(const Module &m, std::vector<Instruction *> &toDelete,
+                        CallInst *curCallInstruction, const CallSite *curCall,
+                        Function *indirectFunction)
+{
+  auto indirectCallHandler =
+      indirectCallFunctions.find(indirectFunction->getName());
+
+  if (indirectCallHandler != indirectCallFunctions.end())
+    indirectCallHandler->second(m, toDelete, curCallInstruction, curCall,
+                                indirectFunction);
+}
+
+/// Check indirect calls in the given module, and handle then.
+void taffo::manageIndirectCalls(llvm::Module &m)
```

Comment:
Not the most useful of comments...

In this cpp file it is fine, perhaps we can refer to `handleIndirectCall` for more details.
However, exposed header documentation should be a bit more detailed.

Reply:
I referenced `handleIndirectCall` in the cpp and added more details in the header documentation with [this commit](https://github.com/HEAPLab/TAFFO-submodule-init/pull/4/commits/6737138f192df814f559d394f74739cc80ac91b4)